### DISCUSSION TITLE
FAST API memory leak issue

### DISCUSSION THREAD

**Discussion by zhangyukun230** [fastapi/fastapi#11761] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.websocket("/ws/{client_id}")
async def websocket_endpoint(websocket: WebSocket, client_id: str):
    await websocket.accept()
    user_agent,user_memory = init_user_sever_agent()
    request_word_dir = local_directory + '/' + client_id
    append_files = []
    history:List= []
    try:
        while True:
            try:
                data = await websocket.receive_text()
            except WebSocketDisconnect:
                logging.info(f"Client {client_id} disconnected.")
                break
            message = json.loads(data)
            content = ''
            # 只有文本内容传输
            if message['type'] == 'text':
                content = message['content']

            # 文本内容和文件一起传输。 文件分块上传。
            elif message['type'] == 'file_chunk':
                chunkMessage = message['chunkMessage']
                if chunkMessage['content'] == '<END>':
                    content = message['content']
                else:
                    file_name = chunkMessage['fileName']
                    chunk_index = chunkMessage['chunkIndex']
                    total_chunks = chunkMessage['totalChunks']
                    chunk_content = base64.b64decode(chunkMessage['content'])

                    file_dir = os.path.join(request_word_dir, client_id + '_' + file_name)
                    os.makedirs(os.path.dirname(file_dir), exist_ok=True)
                    chunk_file_path = f"{file_dir}.part{chunk_index}"
                    with open(chunk_file_path, "wb") as f:
                        f.write(chunk_content)

                    # 检查是否所有分片都已接收完毕
                    if all(os.path.exists(f"{file_dir}.part{idx}") for idx in range(total_chunks)):
                        with open(file_dir, "wb") as final_file:
                            for idx in range(total_chunks):
                                chunk_file_path = f"{file_dir}.part{idx}"
                                with open(chunk_file_path, "rb") as chunk_file:
                                    shutil.copyfileobj(chunk_file, final_file)
                                os.remove(chunk_file_path)
                        print("文件的目录是: " + file_dir)
                        append_files.append(file_dir)

            if content == '':
                continue
            content_result = ''
            async for response in getLLMRequest(user_agent, content, request_word_dir, append_files, history):
                content_result += response
                await websocket.send_text(response)
                await asyncio.sleep(0)
            print(content_result)
            await websocket.send_text("<END>")
            append_files = []

    except Exception as e:
        logging.error(f"Error for client {client_id}: {e}")
    finally:
        logging.info(f"Client {client_id} disconnected. Agent removed.")
        del user_memory
        del user_agent
        gc.collect()


if __name__ == "__main__":
    import uvicorn

    logging.basicConfig(level=logging.INFO)
    uvicorn.run(app, host="0.0.0.0", port=80)
```

### Description

Centos machine, Python 3.10
Using Fastapis to build websocket services,. Open some web pages, VIRT and RES increase a lot first, and after closing, RES memory only decreases slightly. But after opening a lot of web pages and then closing them, RES memory will decrease significantly, while VIRT will still increase significantly
This issue does not exist on Mac.
Why is this?
linux
![image](https://github.com/tiangolo/fastapi/assets/37434447/3d4e7c3a-deb9-4450-8cf2-4784f1566326)
mac
![image](https://github.com/tiangolo/fastapi/assets/37434447/e329e714-df84-4251-9aff-70915b585453)

### Operating System

Linux

### Operating System Details

Centos7.2 machine, Python 3.10  fast api 0.111.0

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.10

**Comment by CrasCris** [comment]
Did you solve it ? , i have a api rest api, i just make what to endpoints does in a websocket, but even when i close the websocket i got a lot of memory

**Reply by zhangyukun230** [reply]
@CrasCris  sorry, not solve it.
